src/dex/balancer-v2/LinearMath.ts

```diff
@@ -0,0 +1,285 @@
+import { MathSol } from './balancer-v2-math';
+
+/*
+Note when comparing vs smart contract:
+Math.divDown === MathSol.divDown (i.e. no fixed point)
+uint256.divDown === MathSol.divDownFixed (i.e. has fixed point)
+Similar for mul
+*/
+
+type Params = {
+  fee: bigint;
+  lowerTarget: bigint;
+  upperTarget: bigint;
+};
+
+export function _calcBptOutPerMainIn(
+  mainIn: bigint,
+  mainBalance: bigint,
+  wrappedBalance: bigint,
+  bptSupply: bigint,
+  params: Params,
+): bigint {
+  // Amount out, so we round down overall.
+
+  if (bptSupply == BigInt(0)) {
+    return _toNominal(mainIn, params);
+  }
+
+  const previousNominalMain = _toNominal(mainBalance, params);
+  const afterNominalMain = _toNominal(mainBalance + mainIn, params);
+  const deltaNominalMain = afterNominalMain - previousNominalMain;
+  const invariant = _calcInvariant(previousNominalMain, wrappedBalance);
+  return MathSol.divDown(MathSol.mul(bptSupply, deltaNominalMain), invariant);
+}
+
+export function _calcBptInPerMainOut(
+  mainOut: bigint,
+  mainBalance: bigint,
+  wrappedBalance: bigint,
+  bptSupply: bigint,
+  params: Params,
+): bigint {
+  // Amount in, so we round up overall.
+  const previousNominalMain = _toNominal(mainBalance, params);
+  const afterNominalMain = _toNominal(mainBalance - mainOut, params);
+  const deltaNominalMain = previousNominalMain - afterNominalMain;
+  const invariant = _calcInvariant(previousNominalMain, wrappedBalance);
+  return MathSol.divUp(MathSol.mul(bptSupply, deltaNominalMain), invariant);
+}
+
+export function _calcBptInPerWrappedOut(
+  wrappedOut: bigint,
+  mainBalance: bigint,
+  wrappedBalance: bigint,
+  bptSupply: bigint,
+  params: Params,
+): bigint {
+  // Amount in, so we round up overall.
+  const nominalMain = _toNominal(mainBalance, params);
+  const previousInvariant = _calcInvariant(nominalMain, wrappedBalance);
+  const newWrappedBalance = wrappedBalance - wrappedOut;
+  const newInvariant = _calcInvariant(nominalMain, newWrappedBalance);
+  const newBptBalance = MathSol.divDown(
+    MathSol.mul(bptSupply, newInvariant),
+    previousInvariant,
+  );
+  return bptSupply - newBptBalance;
+}
+
+export function _calcWrappedOutPerMainIn(
+  mainIn: bigint,
+  mainBalance: bigint,
+  params: Params,
+): bigint {
+  // Amount out, so we round down overall.
+  const previousNominalMain = _toNominal(mainBalance, params);
+  const afterNominalMain = _toNominal(mainBalance + mainIn, params);
+  return afterNominalMain - previousNominalMain;
+}
+
+export function _calcWrappedInPerMainOut(
+  mainOut: bigint,
+  mainBalance: bigint,
+  params: Params,
+): bigint {
+  // Amount in, so we round up overall.
+  const previousNominalMain = _toNominal(mainBalance, params);
+  const afterNominalMain = _toNominal(mainBalance - mainOut, params);
```

Comment:
```suggestion
  const afterNominalMain = _toNominal(mainBalance + mainOut, params);
```
As per https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/pool-linear/contracts/LinearMath.sol#L102